src/daft-catalog/src/impls/memory.rs

```diff
@@ -0,0 +1,368 @@
+        let table_name = ident.name();
+
+        Ok((namespace, table_name))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct MemoryTable {
+    name: String,
+    info: Arc<RwLock<InMemoryInfo>>,
+}
+
+impl MemoryTable {
+    pub fn new(name: String, schema: SchemaRef) -> DaftResult<Self> {
+        let pset = Arc::new(MicroPartitionSet::empty());
+
+        let cache_entry = daft_context::partition_cache::put_partition_set_into_cache(pset)?;
+        let cache_key = cache_entry.key();
+
+        let info = InMemoryInfo::new(schema, cache_key, Some(cache_entry), 0, 0, 0, None, None);
+
+        Ok(Self {
+            name,
+            info: Arc::new(RwLock::new(info)),
+        })
+    }
+}
+
+impl Catalog for MemoryCatalog {
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn create_namespace(&self, ident: &Identifier) -> CatalogResult<()> {
+        if ident.has_qualifier() {
+            return Err(CatalogError::unsupported(
+                "MemoryCatalog does not support nested namespaces",
+            ));
+        }
+
+        let namespace = Some(ident.name().to_string());
+
+        if self.tables.read().unwrap().contains_key(&namespace) {
+            return Err(CatalogError::obj_already_exists("namespace", ident));
+        }
+
+        self.tables
+            .write()
+            .unwrap()
+            .insert(namespace, IndexMap::new());
+
+        Ok(())
+    }
+
+    fn create_table(&self, ident: &Identifier, schema: SchemaRef) -> CatalogResult<TableRef> {
+        let (namespace, table_name) = Self::split_table_ident(ident)?;
+
+        {
+            let tables = self.tables.read().unwrap();
+
+            let Some(namespace_tables) = tables.get(&namespace) else {
+                return Err(CatalogError::ObjectNotFound {
+                    type_: "namespace".to_string(),
+                    ident: namespace.unwrap(),
+                });
+            };
+
+            if namespace_tables.contains_key(table_name) {
+                return Err(CatalogError::obj_already_exists("table", ident));
+            }
+        }
+
+        let table = Arc::new(MemoryTable::new(table_name.to_string(), schema)?);
+
+        self.tables
+            .write()
+            .unwrap()
+            .get_mut(&namespace)
+            .unwrap()
+            .insert(table_name.to_string(), table.clone());
+
+        Ok(table)
+    }
+
+    fn drop_namespace(&self, ident: &Identifier) -> CatalogResult<()> {
+        if ident.has_qualifier() {
+            return Err(CatalogError::obj_not_found("namespace", ident));
+        }
+
+        let namespace = Some(ident.name().to_string());
+
+        match self.tables.write().unwrap().shift_remove(&namespace) {
+            Some(_) => Ok(()),
+            None => Err(CatalogError::obj_not_found("namespace", ident)),
+        }
+    }
+
+    fn drop_table(&self, ident: &Identifier) -> CatalogResult<()> {
+        let (namespace, table_name) = Self::split_table_ident(ident)?;
```

Comment:
Instead of split_table_ident, you could use the .qualifier and .drop on Identifier. Also would make sense to add `.split(&self)` to the Identifier type itself to give us this `(qualifier, name)` pair because this pattern is used elsewhere. I had to do similar in the session's table name resolution for reference.

Reply:
I'm doing this instead because I also want to check for nested namespaces